File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        i, j = 0, 0
        freq_map = defaultdict(int)
        maxi = 0
        while j < len(s):
            freq_map[s[j]] += 1
            
            while (j-i+1) - max(freq_map.values()) > k:
                freq_map[s[i]] -= 1
                i += 1
            maxi = max(maxi, j-i+1)
            j += 1
        return maxi

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int i = 0;
        int j = 0;
        int currFreq;
        int maxi = 0;
        Map<Character, Integer> freqMap = new HashMap<>();
        while(j < s.length()) {
            currFreq = freqMap.getOrDefault(s.charAt(j), 0);
            freqMap.put(s.charAt(j), ++currFreq);
            
            while((j-i+1)-Collections.max(freqMap.values())>k) {
                currFreq = freqMap.get(s.charAt(i));
                freqMap.put(s.charAt(i), --currFreq);
                ++i;
            }
            maxi = Math.max(maxi, j-i+1);
            ++j;
        }
        return maxi;
        
    }
}